content/en/blog/_posts/data-tbd/index.md

```diff
@@ -0,0 +1,284 @@
+reserved or not even allocated. It may be possible to [remove this
+limitation](https://github.com/kubernetes/kubernetes/issues/114005) in the
+future.
+
+## Writing a resource driver
+
+A dynamic resource allocation driver consist of two separate-but-coordinating
+components: a centralized controller and a daemonset of node-local kubelet
+plugins. Most of the work required by the centralized controller to coordinate
+with the scheduler can be handled by boilerplate code. Only the business logic
+required to actually allocate ResourceClaims against the ResourceClasses owned
+by the plugin needs to be customized. As such, the following package is
+provide by Kubernetes to include APIs for invoking this boilerplate code as
+well as a Driver interface that one can implement to provide their custom
+business logic:
+
+- [k8s.io/dynamic-resource-allocation/controller](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/controller)
+
+Likewise, boilerplate code can be used to register the node-local plugin with
+the kubelet, as well as start a gRPC server to implement the kubelet plugin
+API. The following package is provided for this purpose:
+
+- [k8s.io/dynamic-resource-allocation/kubeletplugin](https://github.com/kubernetes/dynamic-resource-allocation/tree/master/kubeletplugin)
+
+It is up to the driver developer to decide how these two components
+communicate. The KEP outlines an [approach using
+CRDs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3063-dynamic-resource-allocation#implementing-a-plugin-for-node-resources).
+
+We also plan to provide a complete [example
+driver](https://github.com/kubernetes-sigs/dra-example-driver) that can serve
+as template for other drivers.
+
+## Running the test driver
+
+The following steps bring up a local, one-node cluster directly from the
+Kubernetes source code. Inside the root directory of Kubernetes, run:
+
+```console
+$ hack/install-etcd.sh
+...
+
+$ RUNTIME_CONFIG=resource.k8s.io/v1alpha1 FEATURE_GATES=DynamicResourceAllocation=true DNS_ADDON="coredns" CGROUP_DRIVER=systemd CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock LOG_LEVEL=6 ENABLE_CSI_SNAPSHOTTER=false API_SECURE_PORT=6444 ALLOW_PRIVILEGED=1 PATH=$(pwd)/third_party/etcd:$PATH ./hack/local-up-cluster.sh -O
+...
+To start using your cluster, you can open up another terminal/tab and run:
+
+  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
+...
+```
+
+A recent CRI-O release must be installed. Once the cluster is up, in another
+terminal run the test driver controller. `KUBECONFIG` must be set for all of
+the following commands.
+
+```console
+$ go run ./test/e2e/dra/test-driver --feature-gates ContextualLogging=true -v=5 controller
+```
+
+In another terminal, run the kubelet plugin:
+
+```console
+$ sudo mkdir -p /var/run/cdi && sudo chmod a+rwx /var/run/cdi /var/lib/kubelet/plugins_registry /var/lib/kubelet/plugins/
```

Comment:
This could be part of the "setup" script from above.

Independently, are these three necessary? I have never needed to do this in my environment:
```
sudo chmod a+rwx /var/run/cdi
sudo chmod a+rwx /var/lib/kubelet/plugins_registry
sudo chmod a+rwx /var/lib/kubelet/plugins/
```

If those are not actually necessary, then the whole setup script would be:
```
$ cat run-k8s.sh
#!/usr/bin/env bash

sudo rm -rf /tmp/kube*
sudo rm -rf /var/run/kubernetes

sudo mkdir -p /var/run/cdi

export RUNTIME_CONFIG=resource.k8s.io/v1alpha1
export FEATURE_GATES=DynamicResourceAllocation=true 
export DNS_ADDON="coredns" 
export CGROUP_DRIVER=systemd 
export CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/crio/crio.sock 
export LOG_LEVEL=6 
export ENABLE_CSI_SNAPSHOTTER=false 
export API_SECURE_PORT=6444 
export ALLOW_PRIVILEGED=1 
export PATH=$(pwd)/third_party/etcd:$PATH
./hack/local-up-cluster.sh -O
```

Where you then run:
```console
$ hack/install-etcd.sh
...
$ ./run-k8s.sh
```

Reply:
> I have never needed to do this in my environment:|

Are you running the "go" command with the kubelet plugin as root? Then it is not needed. But when using a non-privileged user account it is necessary.

Reply:
I've never tried to run things exactly as you described here. I always deploy privileged pods, so that probably explains it.

Reply:
That said, would it make sense to just have them run the `go` command as `sudo` so as not to corrupt their environment for future "real" deployments of the components as pods.

Reply:
When running "sudo go", it will not use the already populated Go cache of the current user. Let me add that as explanation and also point out that permissions should get restored when done.